### PR TITLE
feat(db/common/boltdb) shallow bucket hierarchies

### PR DESCRIPTION
**Pull request by shino**
Only "db add" works (or even it does not work)

### CODE REVIEW

**Comment by shino** [comment]
Preliminary measurement shows 500MB db file size for mitre v5 data set.

```
% CGO_ENABLED=0 /usr/bin/time -v go run cmd/vuls/main.go db add ~/extracted/vuls-data-extracted-mitre-v5  --dbpath ./vuls.lo.db 2>&1 L
2024/07/26 18:41:37 INFO Get Metadata
2024/07/26 18:41:37 INFO Put Vulnerability Data
2024/07/26 18:44:11 INFO Put DataSource
2024/07/26 18:44:11 INFO Put Metadata
        Command being timed: "go run cmd/vuls/main.go db add /home/shino/extracted/vuls-data-extracted-mitre-v5 --dbpath ./vuls.lo.db"
        User time (seconds): 111.32
        System time (seconds): 15.97
        Percent of CPU this job got: 82%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:34.24
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 1164336
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 6482
        Minor (reclaiming a frame) page faults: 385238
        Voluntary context switches: 387459
        Involuntary context switches: 427
        Swaps: 0
        File system inputs: 2149632
        File system outputs: 1029944
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
% ll vuls.lo.db
-rw------- 1 shino shino 494M Jul 26 18:44 vuls.lo.db
```

File: pkg/db/common/boltdb/boltdb.go [hunk 1]
```diff
@@ -573,57 +580,54 @@ func (c *Connection) PutVulnerabilityData(root string) error {
...
+		if err != nil {
+			return errors.Wrapf(err, "create bucket:%q if not exists", d.Ecosystem)
+		}
+		edb, err := eb.CreateBucketIfNotExists([]byte("detection"))
+		if err != nil {
+			return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("%s:detection", d.Ecosystem))
+		}
+
+		criteriaMap := map[detectionTypes.Ecosystem]criteriaTypes.Criteria{}
+		if bs := edb.Get([]byte(data.ID)); bs != nil {
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:594]:
```suggestion
		if bs := edb.Get([]byte(data.ID)); len(bs) > 0 {
```

File: pkg/db/common/boltdb/boltdb.go [hunk 2]
```diff
@@ -573,57 +580,54 @@ func (c *Connection) PutVulnerabilityData(root string) error {
...
-			epb, err := eb.CreateBucketIfNotExists([]byte(p))
-			if err != nil {
-				return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("%s:%s", d.Ecosystem, p))
-			}
-
-			eprb, err := epb.CreateBucketIfNotExists([]byte(data.ID))
-			if err != nil {
-				return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("%s:%s:%s", d.Ecosystem, p, data.ID))
+			var rootIDs []string
+			if bs := eib.Get([]byte(p)); bs != nil {
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:619]:
```suggestion
			if bs := eib.Get([]byte(p)); len(bs) > 0 {
```

File: pkg/db/common/boltdb/boltdb.go [hunk 3]
```diff
@@ -670,23 +674,20 @@ func putAdvisory(tx *bolt.Tx, data dataTypes.Data, roots map[string]dbTypes.Vuln
...
-		vaab, err := vab.CreateBucketIfNotExists([]byte(a.Content.ID))
-		if err != nil {
-			return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("vulnerability:advisory:%s", a.Content.ID))
-		}
-
-		vaasb, err := vaab.CreateBucketIfNotExists([]byte(data.DataSource))
-		if err != nil {
-			return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("vulnerability:advisory:%s:%s", a.Content.ID, data.DataSource))
+		aMap := map[string]advisoryTypes.Advisory{}
+		if bs := vab.Get([]byte(a.Content.ID)); bs != nil {
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:678]:
```suggestion
		if bs := vab.Get([]byte(a.Content.ID)); len(bs) > 0 {
```

File: pkg/db/common/boltdb/boltdb.go [hunk 4]
```diff
@@ -708,23 +709,21 @@ func putVulnerability(tx *bolt.Tx, data dataTypes.Data, roots map[string]dbTypes
...
-		vvvb, err := vvb.CreateBucketIfNotExists([]byte(v.Content.ID))
-		if err != nil {
-			return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("vulnerability:vulnerability:%s", v.Content.ID))
-		}
 
-		vvvsb, err := vvvb.CreateBucketIfNotExists([]byte(data.DataSource))
-		if err != nil {
-			return errors.Wrapf(err, "create bucket:%q if not exists", fmt.Sprintf("vulnerability:vulnerability:%s:%s", v.Content.ID, data.DataSource))
+		vMap := map[string]vulnerabilityTypes.Vulnerability{}
+		if bs := vvb.Get([]byte(v.Content.ID)); bs != nil {
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:714]:
```suggestion
		if bs := vvb.Get([]byte(v.Content.ID)); len(bs) > 0 {
```

File: pkg/db/common/boltdb/boltdb.go (hunk 1 above)
Comment (MaineK00n) [reply pkg/db/common/boltdb/boltdb.go:594]: https://go.dev/play/p/Gfit7rob8y-

File: pkg/db/common/boltdb/boltdb.go [hunk 5]
```diff
@@ -536,9 +536,16 @@ func (c *Connection) PutVulnerabilityData(root string) error {
 				return errors.Wrapf(err, "decode %s", path)
 			}
 
+			es := []string{}
```
Comment (MaineK00n) [inline pkg/db/common/boltdb/boltdb.go:539]:
```suggestion
			es := []detectionTypes.Detection{}
```

**Comment by MaineK00n** [comment]
```console
$ /usr/bin/time -v ./vuls db add ~/Downloads/vuls-data-extracted-mitre-v5
2024/07/29 19:38:14 INFO Get Metadata
2024/07/29 19:38:14 INFO Put Vulnerability Data
2024/07/29 19:38:22 INFO Put DataSource
2024/07/29 19:38:22 INFO Put Metadata
	Command being timed: "./vuls db add /home/mainek00n/Downloads/vuls-data-extracted-mitre-v5"
	User time (seconds): 8.85
	System time (seconds): 1.49
	Percent of CPU this job got: 126%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:08.19
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 1143568
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 295005
	Voluntary context switches: 5022
	Involuntary context switches: 986
	Swaps: 0
	File system inputs: 0
	File system outputs: 978280
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0

$ du -h ~/.cache/vuls/*
478M	/home/mainek00n/.cache/vuls/vuls.db
```

**Review by MaineK00n — APPROVED** [review]

**Pull request merged**